### ISSUE TITLE
ec2_asg module sporadically fails to get async notification

### ISSUE THREAD

**Issue by ansibot** [ansible/ansible#29801]
From @srkiNZ84 on 2016-05-09T22:40:20Z
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

ec2_asg
##### ANSIBLE VERSION

```
ansible 2.0.0.2
```
##### OS / ENVIRONMENT

Running from: Ubuntu Linux
Managing: Ubuntu Linux
##### SUMMARY

We (seeming randomly) get cases where we'll do a blue/green deploy using the "ec2_asg" module and have an async task waiting for the module to return a result. The task that waits for the result never gets the async notification and therefore fails despite the deploy succeeding.
##### STEPS TO REPRODUCE
- Create a new launch config (our new "blue" deploy)
- Run the "ec2_asg" task with the new launch config (with async set and "poll: 0")
- Have a task later in the playbook waiting on the result
- Confirm that the deploy succeeds in AWS (new instances brought up, old ones terminated)
- See that the "async_status" job never gets the notification that the deploy has happened

```
- name: Create integration tier launch configuration
  ec2_lc:
    name: "{{ environ }}-int-launch-config-{{ current_time }}"
    [OMITTED FOR BREVITY]
  register: int_launch_configuration

- name: Create Integration Autoscaling group
  ec2_asg:
    name: "{{ environ }}-int-asg"
    launch_config_name: "{{ environ }}-int-launch-config-{{ current_time }}"
    vpc_zone_identifier: "{{ int_subnets }}"
    health_check_type: "ELB"
    health_check_period: 400
    termination_policies: "OldestInstance"
    replace_all_instances: yes
    wait_timeout: 2400
    replace_batch_size: "{{ int_replace_batch_size }}"
  async: 1000
  poll: 0
  register: int_asg_sleeper

  - name: 'int ASG - check on fire and forget task'
    async_status: jid={{ int_asg_sleeper.ansible_job_id }}
    register: int_asg_job_result
    until: int_asg_job_result.finished
    retries: 60
    delay: 15
```
##### EXPECTED RESULTS

Expected that when the deploy succeeds and the "old" instances are terminated, the Async job gets the message and reports success.
##### ACTUAL RESULTS

It appears that the "file" mechanism which Python/Ansible use for checking on the status of background jobs fails and the file is never populated, despite the job finishing. Therefore the job polling the file times out eventually.

```
08:03:34.063 TASK [launch-config : int ASG - check on fire and forget task] *****************
08:03:34.130 fatal: [localhost]: FAILED! => {"failed": true, "msg": "ERROR! The conditional check 'int_asg_job_result.finished' failed. The error was: ERROR! error while evaluating conditional (int_asg_job_result.finished): ERROR! 'dict object' has no attribute 'finished'"}
```

Copied from original issue: ansible/ansible-modules-core#3625

**Comment by ansibot** [comment]
From @srkiNZ84 on 2016-05-09T22:40:20Z

I should also mention that previously we didn't do this using the "async" mechanism, rather we did it synchronously and didn't see this issue.
The problem started happening when we implemented async in order to deploy to three auto scaling groups in parallel.

**Comment by ansibot** [comment]
From @electrofelix on 2016-05-09T22:40:20Z

Notice that your task is not timing out, instead it's hitting an exception about the `finished` field not existing. This is because initially that field is not defined until the task is complete.

I suggest changing the until to use the following format and retest:
```
  until: int_asg_job_result.finished is defined and int_asg_job_result.finished
```

The need to do this has possibly has been fixed by https://github.com/ansible/ansible-modules-core/commit/b75f305079c89ba44358890db7145dd507b3dc92, though it could still occur if the task hit an exception since the json written out in those cases do not contain the 'finished' field.

I believe the problem stems from the initial json written out would have been:
```
{ "started" : 1, "ansible_job_id" : jid, "results_file" : job_path }
```
followed by:
```
{ "started" : 1, "ansible_job_id" : jid }
```

Notice there is no finish field there, so the until above would not see 'finished' if it checked too soon.

**Comment by ansibot** [comment]
From @electrofelix on 2016-05-09T22:40:20Z

Think this might be a duplicate of #2374

**Comment by ansibot** [comment]
cc @garethr
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by biomassives** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (110/100 (110%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext3.iad.github.net ([192.30.252.194]:43906 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89)
	(envelope-from <noreply@github.com>)
	id 1drZEt-001otn-75
	for greg@biomassiv.es; Mon, 11 Sep 2017 19:39:43 -0500
Date: Mon, 11 Sep 2017 17:39:03 -0700
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1505176743;
	bh=vMXfKCKqjOfCVRPX4+xsbE++nKMBrSSRU4p7xyTlEwE=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=STbZ+UYaHexm0YsYx2TxXzqJIKi7GXNTFHp7k9FrpqcmQZjNn4U1cCLrS9Yk6AZTd
	 tStOjet+SWR5BVyaFEarXG+2P/rCZSMAmbrsgc4BZHCgntUmJQLA2IBmHVpk9UZLVZ
	 l2luM7yswfTexN8gqnAo/PnPhzIu8RDZlt7zUc0g=

**Comment by ansibot** [comment]
cc @ryansb @s-hertel
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @willthames
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
cc @jillr @tremble
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by electrofelix** [comment]
@jillr @tremble I would suggest that this issue has been resolved based on some of my comments above, and can be worked around for the edge cases.

These days if the task is successfully started using async https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/utilities/logic/async_wrapper.py#L270-L271 should ensure the above error doesn't appear.

Going to unsubscribe from this now.

**Comment by s-hertel** [comment]
Closing as per above
